### PR TITLE
[CalyxToFSM] Add CalyxToFSM conversion pass

### PR DESCRIPTION
**Pull request by mortbopet**
This commit represents a lowering pass for converting a Calyx control schedule (seq/if/while/enable) into an FSM representation. The FSM is embedded within the Calyx component, and references both the group symbols and cell SSA values.

The lowering method is a fairly straightforward conversion, which leaves plenty of canonicalization opportunities to remove redundant states in the generated FSM. However, doing it as presented in this PR lends itself to some very clean code (which I prefer, rather than prematurely optimizing during lowering) as well as meaningful naming of states (which i think is fairly important).

### CODE REVIEW

**Comment by mortbopet** [comment]
clang-tidy complains about recursive call, can be ignored.

**Comment by mortbopet** [comment]
tidy can be ignored (recursive calls) - prefer not to add NO-LINT's since almost all functions would need it to fully silence the warning.

File: lib/Conversion/CalyxToFSM/CalyxToFSM.cpp [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+    // Recurse into the body of the branch, with an exit state targeting
+    // 'nextState'.
+    if (failed(dispatch(branchStateOp, innerBranchOp, nextState)))
+      return failure();
+    return success();
+  };
+
+  // Then branch.
+  if (failed(lowerBranch(ifOp.cond(), "then", false,
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/CalyxToFSM.cpp:151]: Nit: `/*invert=*/` here and below

File: lib/Conversion/CalyxToFSM/CalyxToFSM.cpp [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+  /// Generates a new state name based on the current state stack and the
+  /// provided suffix. The new suffix is pushed onto the state stack. Returns a
+  /// guard object which pops the new suffix upon destruction.
+  StateScopeGuard pushStateScope(StringRef suffix) {
+    std::string name;
+    llvm::raw_string_ostream ss(name);
+    llvm::interleave(
+        stateStack, ss, [&](const auto &it) { ss << it; }, "_");
+    name += +"_" + suffix.str();
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/CalyxToFSM.cpp:98]: Is this a typo?

```cpp
name += +"_" + suffix.str();
        ^
```

**Review by cgyurgyik — COMMENTED** [review]
This also looks pretty straight-forward, I will need to spend a bit more time looking through the implementation.

**Comment by mortbopet** [comment]
@cgyurgyik, @mikeurbach bumping this due to the recent landing of FSMToSV support. This, #3214 + a few minor passes will have us doing Calyx end-to-end in CIRCT.

**Comment by mikeurbach** [comment]
Thanks for the heads up, I was just wondering what the status was. I will take another look.

**Review by cgyurgyik — APPROVED** [review]
In general LGTM.

File: test/Conversion/CalyxToFSM/lower-if.mlir [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          }
+        } else {
+          calyx.seq {
+            calyx.enable @false
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (mikeurbach) [inline test/Conversion/CalyxToFSM/lower-if.mlir:101]: uber nit: new line

**Review by mikeurbach — APPROVED** [review]
I think the machinery of the pass looks good. The individual visitor methods look good to me, very nicely done. From what I recall of tdcc in the Rust compiler, this seems familiar. At some point I'll take a deeper look, and think about whether we can add further tweaks to optimize out extra states, take advantage of static timing, etc., but those can easily be added on top of this baseline. Thanks for pushing on this!

**Pull request merged**